### PR TITLE
Adds support for multi-line macro inputs. Makes MacroUtils more modular. Adds rawTest macro. Fix #1524.

### LINKED ISSUE #1524: Move rawTest macro to core
`rawTest:HomePage[]` should be rendered
```
[source,groovy]
.src/test/groovy/example/micronaut/HomePage.groovy
----
include::{sourceDir}/micronaut-security-jwt-cookie/micronaut-security-jwt-cookie-maven-kotlin/src/test/groovy/example/micronaut/HomePage.groovy[]
----
```

### PR DESCRIPTION
**Pull request by AndreaLaGrotteria**

### CODE REVIEW

File: buildSrc/src/main/java/io/micronaut/guides/core/MacroUtils.java [hunk 1]
```diff
@@ -44,7 +57,9 @@ public static List<String> addIncludes(@NonNull GuidesOption option,
             }
         } else {
             List<String> attributes = new ArrayList<>();
-            attributes.add("lines=" + licenseLoader.getNumberOfLines() + "..-1");
+            if (licenseLoader != null) {
```
Comment (sdelamo) [inline buildSrc/src/main/java/io/micronaut/guides/core/MacroUtils.java]: why is `licenseLoader` @Nullable? it should not be.

File: buildSrc/src/main/java/io/micronaut/guides/core/MacroUtils.java [hunk 2]
```diff
@@ -20,16 +20,29 @@ public static String getSourceDir(@NonNull String slug, @NonNull GuidesOption op
...
+     * @param indent        The indent value for the include directive. If empty no indent is added.
+     * @param tags          The list of tags for the include directive.
+     * @return A list of lines with include directives.
+     */
     @NonNull
     public static List<String> addIncludes(@NonNull GuidesOption option,
                                            @NonNull String slug,
                                            @NonNull String sourcePath,
-                                           @NonNull LicenseLoader licenseLoader,
+                                           LicenseLoader licenseLoader,
```
Comment (sdelamo) [inline buildSrc/src/main/java/io/micronaut/guides/core/MacroUtils.java]: If licenseLoader is Nullable it should be annotated with `@Nullable`

File: buildSrc/src/main/java/io/micronaut/guides/core/RawTestMacroSubstitution.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        List<String> tagNames = extractTags(str);
+        List<String> tags = (tagNames != null && !tagNames.isEmpty())
+                ? tagNames.stream().map(it -> "tag=" + it).toList()
+                : Collections.emptyList();
+
+        String indent = "";
+
+        String sourcePath = rawTestPath(guidesConfiguration, appName, name, option);
+
+        List<String> lines = addIncludes(option, slug, sourcePath, null, option.getTestFramework().toTestFramework().getDefaultLanguage().getExtension(),indent, tags);
```
Comment (sdelamo) [inline buildSrc/src/main/java/io/micronaut/guides/core/RawTestMacroSubstitution.java]: supply `licenseLoader`

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Review by sdelamo — CHANGES_REQUESTED** [review]
test fail

File: buildSrc/src/test/java/io/micronaut/guides/core/RawTestMacroSubstitutionTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        String expectedGroovy = """
+                [source,groovy]
+                .src/test/groovy/example/micronaut/HomePage.groovy
+                ----
+                include::{sourceDir}/micronaut-security-jwt-cookie/micronaut-security-jwt-cookie-gradle-groovy/src/test/groovy/example/micronaut/HomePage.groovy[lines=16..-1]
+                ----
+                """;
+        assertEquals(expectedGroovy, resGroovy);
+    }
+}
```
Comment (sdelamo) [inline buildSrc/src/test/java/io/micronaut/guides/core/RawTestMacroSubstitutionTest.java]: The following test should pass. **Every macro needs to support multi line strings.** 

```suggestion
package io.micronaut.guides.core;

import io.micronaut.starter.api.TestFramework;
import io.micronaut.starter.options.BuildTool;
import io.micronaut.starter.options.Language;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

@MicronautTest(startApplication = false)
public class RawTestMacroSubstitutionTest {
    @Inject
    private RawTestMacroSubstitution rawTestMacroSubstitution;

    @Test
    void testSubstitute() {
        String str = """
Create three pages:

rawTest:HomePage[]
""";
        String resJava = rawTestMacroSubstitution.substitute(str, "micronaut-security-jwt-cookie", new GuidesOption(BuildTool.GRADLE, Language.JAVA, TestFramework.SPOCK));
        String expectedJava = """
                Create three pages:
                
                [source,groovy]
                .src/test/groovy/example/micronaut/HomePage.groovy
                ----
                include::{sourceDir}/micronaut-security-jwt-cookie/micronaut-security-jwt-cookie-gradle-java/src/test/groovy/example/micronaut/HomePage.groovy[lines=16..-1]
                ----
                """;
        assertEquals(expectedJava, resJava);
        String resKotlin = rawTestMacroSubstitution.substitute(str, "micronaut-security-jwt-cookie", new GuidesOption(BuildTool.GRADLE, Language.KOTLIN, TestFramework.SPOCK));
        String expectedKotlin = """
                Create three pages:
                
                [source,groovy]
                .src/test/groovy/example/micronaut/HomePage.groovy
                ----
                include::{sourceDir}/micronaut-security-jwt-cookie/micronaut-security-jwt-cookie-gradle-kotlin/src/test/groovy/example/micronaut/HomePage.groovy[lines=16..-1]
                ----
                """;
        assertEquals(expectedKotlin, resKotlin);
        String resGroovy = rawTestMacroSubstitution.substitute(str, "micronaut-security-jwt-cookie", new GuidesOption(BuildTool.GRADLE, Language.GROOVY, TestFramework.SPOCK));
        String expectedGroovy = """
                Create three pages:
                
                [source,groovy]
                .src/test/groovy/example/micronaut/HomePage.groovy
                ----
                include::{sourceDir}/micronaut-security-jwt-cookie/micronaut-security-jwt-cookie-gradle-groovy/src/test/groovy/example/micronaut/HomePage.groovy[lines=16..-1]
                ----
                """;
        assertEquals(expectedGroovy, resGroovy);
    }
}
```

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/java/io/micronaut/guides/core/RawTestMacroSubstitution.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    @NonNull
+    private static String rawTestPath(@NonNull GuidesConfiguration guidesConfiguration,
+                              @NonNull String appName,
+                              @NonNull String name,
+                              @NonNull GuidesOption option) {
+        String module = appName.isEmpty() ? "" : appName + "/";
+        String fileExtension = option.getTestFramework().toTestFramework().getDefaultLanguage().getExtension();
+        String langTestFolder = option.getTestFramework().toTestFramework().getDefaultLanguage().getTestSrcDir();
+        return module+langTestFolder + "/" + guidesConfiguration.getPackageName().replace(".", "/") + "/" + name + "." + fileExtension;
```
Comment (sdelamo) [inline buildSrc/src/main/java/io/micronaut/guides/core/RawTestMacroSubstitution.java]:
```suggestion
        return module + langTestFolder + "/" + guidesConfiguration.getPackageName().replace(".", "/") + "/" + name + "." + fileExtension;
```

File: buildSrc/src/test/java/io/micronaut/guides/core/SourceMacroSubstitutionTest.java [hunk 6]
```diff
@@ -18,7 +17,7 @@ class SourceMacroSubstitutionTest {
 
     @Test
     void testSubstitute(){
-        String str = "source:GithubConfiguration[]";
+        String str = "source:GithubConfiguration[]\n";
```
Comment (sdelamo) [inline buildSrc/src/test/java/io/micronaut/guides/core/SourceMacroSubstitutionTest.java:20]: why do you need to add a \n line character. 

```suggestion
        String str = "source:GithubConfiguration[]\n";
```

**Review by sdelamo — APPROVED** [review]

File: buildSrc/src/test/java/io/micronaut/guides/core/SourceMacroSubstitutionTest.java (hunk 6 above)
Comment (sdelamo) [reply buildSrc/src/test/java/io/micronaut/guides/core/SourceMacroSubstitutionTest.java:20]: this looks like a bad smell to me.

Comment (AndreaLaGrotteria) [reply buildSrc/src/test/java/io/micronaut/guides/core/SourceMacroSubstitutionTest.java:20]: If we have multi-line inputs, the different lines will be separated by a \n, so it was to test that. If you look at the other tests you will see that input without \n is also tested and works.

**Pull request merged**